### PR TITLE
[Fortimanager] Switch from `kv` to `script` Processors to Improve Performance

### LINKED ISSUE #8306: [fortinet_fortimanager] `logs-fortinet_fortimanager.log` kv:kv_temp processor is expensive
As of the at least the integration version `2.8.0` the `kv:kv_temp` processor in `logs-fortinet_fortimanager.log` is _relatively_ extremely expensive.

image

The split patterns are somewhat odd, so I'm not sure how they can be optimized, but a review of it probably needs to be done to consider the performance of it.

### PR DESCRIPTION
**Pull request by BenB196**
## Proposed commit message

The KV (`kv_temp`) processor for this pipeline was pretty expensive, #8306, mainly because of the extremely "greedy" `field_split` regex.

This PR redoes the regex for this processor to be more efficient.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] The original regex and the new one are both pretty hard to understand, I made sure that the existing pipeline tests (and all other tests) still passed with these changes. If there are docs that aren't covered by these tests, then they might break, but hard to know for sure.
- [ ] I updated the `value_split` regex as well. I'm not too sure why it was doing a negative lookbehind, the tests pass with it being a standard `=`.

Before these changes the pipeline test would take ~16-19ms:

```ps
elastic-package.exe test pipeline
Run pipeline tests for the package
--- Test results for package: fortinet_fortimanager - START ---
╭───────────────────────┬─────────────┬───────────┬───────────────────────┬────────┬──────────────╮
│ PACKAGE               │ DATA STREAM │ TEST TYPE │ TEST NAME             │ RESULT │ TIME ELAPSED │
├───────────────────────┼─────────────┼───────────┼───────────────────────┼────────┼──────────────┤
│ fortinet_fortimanager │ log         │ pipeline  │ test-fortimanager.log │ PASS   │    18.3125ms │
╰───────────────────────┴─────────────┴───────────┴───────────────────────┴────────┴──────────────╯
--- Test results for package: fortinet_fortimanager - END   ---
Done
```

After these changes the pipeline test takes ~10-13ms:
(Note: This isn't 100% demonstrative of the performance gain here, there is only really one long/nested kv example doc here, and this is where the bulk of the speed up comes from)

```ps
elastic-package.exe test pipeline
Run pipeline tests for the package
--- Test results for package: fortinet_fortimanager - START ---
╭───────────────────────┬─────────────┬───────────┬───────────────────────┬────────┬──────────────╮
│ PACKAGE               │ DATA STREAM │ TEST TYPE │ TEST NAME             │ RESULT │ TIME ELAPSED │
├───────────────────────┼─────────────┼───────────┼───────────────────────┼────────┼──────────────┤
│ fortinet_fortimanager │ log         │ pipeline  │ test-fortimanager.log │ PASS   │    12.2242ms │
╰───────────────────────┴─────────────┴───────────┴───────────────────────┴────────┴──────────────╯
--- Test results for package: fortinet_fortimanager - END   ---
Done
```

The main "win" here is the `field_split` change.

Testing on https://regex101.com/ against: https://github.com/elastic/integrations/blob/3087d3936e9cef025c1420c96f3be7096c164187/packages/fortinet_fortimanager/data_stream/log/_dev/test/pipeline/test-fortimanager.log#L25

For the old regex; `( (?=(?:[^'"]|'[^']*'|"[^"]*")*$))`:
- With the Java 8 engine, we get the error `The Java regex engine reached the maximum call stack size.`
- With the PCRE2 engine, we see that it finds the expected 23 matches, but takes ~34k steps.

For the new regex: `((?<=(?:[a-z0-9\"])) (?=(?:[a-z_]+=)))`:
- With the Java 8 engine, we see it find the expected 23 matches
- With the PCRE2 engine, we see if find the expected 23 matches, but it takes just 400 steps.

To briefly cover the change to `value_split`. Testing against the largest `field_split` value of:
https://github.com/elastic/integrations/blob/3087d3936e9cef025c1420c96f3be7096c164187/packages/fortinet_fortimanager/data_stream/log/_dev/test/pipeline/test-fortimanager.log#L25

For the old regex; `(?<!\\)=`:
- It took 200 steps

With the new regex; `=`:
- It takes 100 steps

## Related issues

- Closes #8306

### CODE REVIEW

File: packages/fortinet_fortimanager/changelog.yml [hunk 1]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: 2.10.1
+  changes:
+    - description: Improve performance of kv processor.
+      type: bugfix
+      link: https://github.com/elastic/integrations/pull/8523
```
Comment (efd6) [inline packages/fortinet_fortimanager/changelog.yml:6]:
```suggestion
- version: 2.11.0
  changes:
    - description: Improve performance of kv processor.
      type: enhancement
      link: https://github.com/elastic/integrations/pull/8523
```

File: packages/fortinet_fortimanager/manifest.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 format_version: "3.0.0"
 name: fortinet_fortimanager
 title: Fortinet FortiManager Logs
-version: "2.10.0"
+version: "2.10.1"
```
Comment (efd6) [inline packages/fortinet_fortimanager/manifest.yml:4]:
```suggestion
version: "2.11.0"
```

File: packages/fortinet_fortimanager/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -29,8 +29,8 @@ processors:
       field: temp
       tag: 'kv_temp'
       target_field: _temp
-      field_split: ( (?=(?:[^'"]|'[^']*'|"[^"]*")*$))
-      value_split: (?<!\\)=
+      field_split: ((?<=(?:[a-z0-9\"])) (?=(?:[a-z_]+=)))
+      value_split: =
       trim_value: "\""
```
Comment (efd6) [inline packages/fortinet_fortimanager/data_stream/log/elasticsearch/ingest_pipeline/default.yml:34]: While we are here...
```suggestion
      trim_value: '"'
```

**Review by efd6 — COMMENTED** [review]
I don't believe this is a safe change.

old:
- split fields on space with ZWLA:(no quotes, or single quoted string with no internal single quotes, or double quoted string with no internal double quotes)
- split value on '=' with ZWLB:(no preceding '\')

new:
- split fields on space with ZWLB:(any alnum or quote) and ZWLA:(one or more alpha or '_' followed by a '=')
- split value on '='

Base on this, the following field appears that it would be incorrectly split between "is" and "this". The fields would then give "label":"value of an equals is" and "this":"".
label='value of an equals is this='

**Comment by efd6** [comment]
> Would you happen to have an example of this from a FortiManager log or are you just stating that this could be a functional change?

This is a constructed example, but any quoted value with a `=` preceded by lable-valid text with a space in front would be split on that space and then have the incorrect field be value-split on the `=`. I could see that text ending up in logs.

**Comment by BenB196** [comment]
Thanks for the clarification here @efd6. Doing some digging, unfortunately on my end I don't keep `event.original` so can't rely on my logs. However, it does look like filebeat contains significantly more log examples; https://github.com/elastic/beats/blob/main/x-pack/filebeat/module/fortinet/fortimanager/test/generated.log, I'll look at implementing some more of these examples into this integration and see if any issues come up.

**Comment by BenB196** [comment]
Hmm, so this would be far more challenging than initially anticipated.

I don't think that the examples in; https://github.com/elastic/beats/blob/main/x-pack/filebeat/module/fortinet/fortimanager/test/generated.log are actually valid, as none of these pass the existing pipeline tests (even with manually adding the `<134>`).

Not sure if there are a way to get more valid examples to build tests with.

However, looking at your concern, you were correct in that were could have been an issue here.

I've pushed a new pattern here:

`((?<=(?:[a-z0-9\"])) (?=(?:[a-z_]+=(?:[0-9]|.[^ ]))))`

```diff
- ((?<=(?:[a-z0-9\"])) (?=(?:[a-z_]+=)))
+ ((?<=(?:[a-z0-9\"])) (?=(?:[a-z_]+=(?:[0-9]|.[^ ]))))
```

The key addition here is: `(?:[0-9]|.[^ ])` as part of the look ahead after looking for the `=`.

This should guard against something like your example:

`label='value of an equals is this='`

While still allowing for things like (key equal to a single digit value):

`euid=1`

I'm not expect anything like (key equal to a single character value):

`key=a`

**Comment by efd6** [comment]
Yes. Using a regular language to parse things that require a stack is always going to be unpleasant at best, and in the general case impossible. I honestly think that what we have now is the best that we can get given the technology available.

**Comment by mrodm** [comment]
Hi @BenB196, please update your branch with the latest contents from main branch. There was an important PR merged updating the CI pipelines. Thanks!

**Comment by BenB196** [comment]
Going to close the PR for now. Still think there is room to improve on the performance of this processor, just don't currently have the time to invest in sorting out the testing and edge-case handling that is required.

**Comment by BenB196** [comment]
@efd6 I've resurrected this PR as I'd like to see what additional things Elastic would like to see to be comfortable with this change.

I recently experienced a cluster ingest backpressure issue where the primary culprit appeared to be a large amount of Fortinet logs processing just slow enough to cause problems.

I applied the changes in this PR to the pipeline and saw the cluster recover and significantly better performance of this pipeline.

Some stats:

30 Minute window from before the change (day before around the same time to get accurate comparison):

image

30 Minutes after the change:

image

After making the change, I also observed no failures on any processors in the pipeline.

**Comment by efd6** [comment]
@BenB196 I'm no longer a code owner on this package, so the decisions here will be down to the owning team. They will git your review ping.

**Comment by taylor-swanson** [comment]
@BenB196, here's another option to consider. For the `fortinet_fortigate` and new `fortinet_fortiproxy` integrations, a `script` processor is used instead of `kv`. Running the pipeline benchmarks yielded these results:

`fortinet_fortimanager` with no changes:
```
--- Benchmark results for package: fortinet_fortimanager - START ---
╭─────────────────────────╮
│ parameters              │
├──────────────────┬──────┤
│ source_doc_count │   26 │
│ doc_count        │ 1000 │
╰──────────────────┴──────╯
╭───────────────────────────╮
│ pipeline_performance      │
├─────────────────┬─────────┤
│ processing_time │   0.22s │
│ eps             │ 4587.16 │
╰─────────────────┴─────────╯
╭────────────────────────────────────╮
│ procs_by_total_time                │
├───────────────────────────┬────────┤
│ kv @ default.yml:28       │ 70.64% │
│ grok @ default.yml:18     │  1.83% │
│ remove @ default.yml:1210 │  0.92% │
│ date @ default.yml:636    │  0.92% │
│ script @ default.yml:1234 │  0.46% │
│ date @ default.yml:646    │  0.46% │
╰───────────────────────────┴────────╯
╭───────────────────────────────────╮
│ procs_by_avg_time_per_doc         │
├───────────────────────────┬───────┤
│ kv @ default.yml:28       │ 154µs │
│ grok @ default.yml:18     │   4µs │
│ remove @ default.yml:1210 │   2µs │
│ date @ default.yml:636    │   2µs │
│ script @ default.yml:1234 │   1µs │
│ date @ default.yml:646    │   1µs │
╰───────────────────────────┴───────╯
```

`fortinet_fortimanager` with the changes from this PR:
```
--- Benchmark results for package: fortinet_fortimanager - START ---
╭─────────────────────────╮
│ parameters              │
├──────────────────┬──────┤
│ source_doc_count │   26 │
│ doc_count        │ 1000 │
╰──────────────────┴──────╯
╭────────────────────────────╮
│ pipeline_performance       │
├─────────────────┬──────────┤
│ processing_time │    0.10s │
│ eps             │ 10204.08 │
╰─────────────────┴──────────╯
╭────────────────────────────────────╮
│ procs_by_total_time                │
├───────────────────────────┬────────┤
│ kv @ default.yml:28       │ 31.63% │
│ grok @ default.yml:18     │  4.08% │
│ remove @ default.yml:1210 │  2.04% │
│ date @ default.yml:646    │  2.04% │
│ date @ default.yml:636    │  2.04% │
│ script @ default.yml:1234 │  1.02% │
╰───────────────────────────┴────────╯
╭──────────────────────────────────╮
│ procs_by_avg_time_per_doc        │
├───────────────────────────┬──────┤
│ kv @ default.yml:28       │ 31µs │
│ grok @ default.yml:18     │  4µs │
│ remove @ default.yml:1210 │  2µs │
│ date @ default.yml:646    │  2µs │
│ date @ default.yml:636    │  2µs │
│ script @ default.yml:1234 │  1µs │
╰───────────────────────────┴──────╯
```

`fortinet_fortimanager` with the script processor:
```
--- Benchmark results for package: fortinet_fortimanager - START ---
╭─────────────────────────╮
│ parameters              │
├──────────────────┬──────┤
│ source_doc_count │   26 │
│ doc_count        │ 1000 │
╰──────────────────┴──────╯
╭────────────────────────────╮
│ pipeline_performance       │
├─────────────────┬──────────┤
│ processing_time │    0.08s │
│ eps             │ 12048.19 │
╰─────────────────┴──────────╯
╭────────────────────────────────────╮
│ procs_by_total_time                │
├───────────────────────────┬────────┤
│ script @ default.yml:39   │ 14.46% │
│ grok @ default.yml:18     │  6.02% │
│ remove @ default.yml:1250 │  2.41% │
│ date @ default.yml:686    │  2.41% │
│ date @ default.yml:676    │  2.41% │
│ script @ default.yml:1274 │  1.20% │
│ date @ default.yml:1049   │  1.20% │
╰───────────────────────────┴────────╯
╭─────────────────────────────────────╮
│ procs_by_avg_time_per_doc           │
├───────────────────────────┬─────────┤
│ script @ default.yml:39   │    12µs │
│ date @ default.yml:1049   │ 6.578µs │
│ grok @ default.yml:18     │     5µs │
│ remove @ default.yml:1250 │     2µs │
│ date @ default.yml:686    │     2µs │
│ date @ default.yml:676    │     2µs │
│ script @ default.yml:1274 │     1µs │
╰───────────────────────────┴─────────╯
```

`fortinet_fortigate`'s script processor:
```
│ script @ default.yml:32  │     9µs │
│ script @ default.yml:77  │     2µs │
```

`fortinet_fortiproxy`'s script processor:
```
│ script @ default.yml:54 │ 16µs │
```

fortigate's implementation looks to be slightly faster, although it is split between two processors. It has also been out for longer and has been tested better. fortiproxy's implementation doesn't work as well if there are odd control characters in the message.

**Comment by BenB196** [comment]
Interesting @taylor-swanson I'll take a look in the next day or so to see how easy it'd be to implement the script processor here instead. Even with the PR as it currently is, this pipeline is still one of our most time consuming, so getting it down further would be a nice "win".

Question, I'm not too familiar with the tooling, how'd you run those benchmarks? That something new?

**Comment by taylor-swanson** [comment]
@BenB196, in theory, the script(s) should be a drop in replacement for the kv processor, you'd just have to update the field it extracts from and the destination where the fields end up.

The benchmarks are a relatively new feature to `elastic-package` (I forget what version they were added). They run like a pipeline test, so just navigate to the package folder and run:

```
elastic-package benchmark pipeline
```

I wish it gave more control over the number of iterations to run and stats in the outputs, but it's better than nothing. It helps to run the benchmarks a few times. I'm not entirely sure how elasticsearch optimizes under the hood, but it seems to a take a few runs before it settles into consistent numbers (perhaps the JIT has to warm up first).

**Comment by BenB196** [comment]
Thanks @taylor-swanson for the tip, I went ahead a tested this out and it does significantly help things.

Stats with previous `kv` regex changes:

image
image

Stats with the `script` processors:

image
image

Still not the greatest performance ratio compared to other pipelines, but definitely _significantly_ better than before.

I've pushed the _minimum_ changes required to make this PR work correctly and updated it to reflect to change.

**Pull request merged**